lib/web_ui/lib/src/engine/platform_dispatcher/app_lifecycle_state.dart

```diff
@@ -0,0 +1,86 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+import 'package:ui/ui.dart' as ui;
+
+import '../../engine.dart';
+
+/// Signature of functions added as a listener to [ui.AppLifecycleState] changes
+typedef AppLifecycleStateListener = void Function(ui.AppLifecycleState state);
+
+/// Determines the [ui.AppLifecycleState].
+abstract class AppLifecycleState {
+  static final AppLifecycleState instance = _BrowserAppLifecycleState();
+
+  ui.AppLifecycleState get appLifecycleState => _appLifecycleState;
+  ui.AppLifecycleState _appLifecycleState = ui.AppLifecycleState.resumed;
+
+  final List<AppLifecycleStateListener> _listeners =
+      <AppLifecycleStateListener>[];
+
+  void addListener(AppLifecycleStateListener listener) {
+    if (_listeners.isEmpty) {
+      activate();
+    }
+    _listeners.add(listener);
+    listener(_appLifecycleState);
+  }
+
+  void removeListener(AppLifecycleStateListener listener) {
+    _listeners.remove(listener);
+    if (_listeners.isEmpty) {
+      deactivate();
+    }
+  }
+
+  @protected
+  void activate();
+
+  @protected
+  void deactivate();
+
+  @visibleForTesting
+  void onAppLifecycleStateChange(ui.AppLifecycleState newState) {
+    if (newState != _appLifecycleState) {
+      _appLifecycleState = newState;
+      for (final AppLifecycleStateListener listener in _listeners) {
+        listener(newState);
+      }
+    }
+  }
+}
+
+/// Determines the [ui.AppLifecycleState] by listening to 'beforeunload',
+/// 'visibilitychange' events on the [DomWindow] / [DomHTMLDocument].
+class _BrowserAppLifecycleState extends AppLifecycleState {
+  @override
+  void activate() {
+    domWindow.addEventListener('beforeunload', _beforeUnloadListener);
```

Comment:
This does have the possibility of making Flutter apps less efficient on Firefox (and probably other browsers). See https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#usage_notes

We might need to hook up a way to know if someone is listening to the application lifecycle state in the framework so that we can only register for this event if someone is listening.

In the meantime, I think this is OK to check in, but you might add a TODO to make this improvement.